lib/kube/proxy/server.go

```diff
@@ -299,9 +307,28 @@ func (t *TLSServer) Serve(listener net.Listener) error {
 	// proxied clusters based on the kube_cluster resources.
 	// This watcher is only started for the kube_service if a resource watcher
 	// is configured.
-	if t.kubeClusterWatcher, err = t.startKubeClusterResourceWatcher(t.closeContext); err != nil {
+	kubeClusterWatcher, err := t.startKubeClusterResourceWatcher(t.closeContext)
+	if err != nil {
 		return trace.Wrap(err)
 	}
+	t.mu.Lock()
+	t.kubeClusterWatcher = kubeClusterWatcher
+	t.mu.Unlock()
+
+	// kubeServerWatcher is used by the kube proxy to watch for changes in the
+	// kubernetes servers of a cluster. Proxy requires it to update the kubeServersMap
+	// which holds the list of kubernetes_services connected to the proxy for a given
+	// kubernetes cluster name. Proxy uses this map to route requests to the correct
+	// kubernetes_service. The servers are kept in memory to avoid making unnecessary
+	// unmarshal calls followed by filtering to improve memory usage.
+	if t.kubeServerWatcher != nil {
```

Comment:
Is there some other concurrent code that could set `t.kubeServerWatcher` to nil after you've just set it above?

Reply:
No. `kubeServerWatcher` is initialized when we create the `TLSServer` as opposed to `kubeClusterWatcher` which is created when `TLSServer.Serve` is called. I cannot move the `kubeClusterWatcher` creation to the same place as `kubeServerWatcher` because it requires the listener address for the heartbeats and it's just available when  `TLSServer.Serve` is called. 

We previously had a race with `t.kubeClusterWatcher` when `Serve` and `Close` were called